WIP : support for image sequence data

teogale opened the pull request.

teogale committed:
test

teogale committed:
change the file __init__.py in order for the user to have acces to the
new classes by doing from neo.core import DesireClasses

change imagesequence, the user no longer need to use the method in each
classes of regionofinterest

started to write unittest for the classe imagesequence and the method of
imagesequence

teogale committed:
added file read_txt

added support to read txt file, isabelle txt file format does only
contain value

teogale reviewed (commented):
isabelle file require to rotate the image in order to have the right orientation 
its not done here,   im just giving the general idea of the format
it should be possible to add an option for the user that make the image rotate

teogale committed:
tif file support

added read_tif.py in rawio to import .tif file from a given directory

teogale committed:
added file read_blk.py

uploading code to read blk file (still in progress)

teogale committed:
add code un read_blk.py
added code to re-organize data (more compatiable with neo imagesequence)
change some loop into comprehension, and made some change for code indentation
added some comment

teogale committed:
modification of read_blk.py

made changes in the code for filetype 13
delete part of the code that where not usefull

teogale committed:
unittests regionofinterest

added little test for the region of interest

teogale committed:
compatibility of read_blk.py
read_blk.py wasn't running on python2 due to the use of
python3 fonction
made changed using struct while reading data  to be able to run it on both

teogale committed:
modification in read_blk.py
improved the use of  struct library

teogale committed:
correction of bug in read_blk.py

apdavison commented on neo/core/imagesequence.py:
```diff
@@ -0,0 +1,76 @@
+from neo.core.baseneo import BaseNeo
+
+
+class ImageSequence(DataObject):
```
I think it would be better to inherit from `BaseSignal` instead of from `DataObject`.

apdavison commented on neo/core/imagesequence.py:
```diff
@@ -0,0 +1,76 @@
+        obj.spatial_scale = spatial_scale
+
+        return obj
+
```
a method `__array_finalize()` is needed here, for when objects are created by slicing, for example (see the implementation of AnalogSignal for an example, and https://docs.scipy.org/doc/numpy/user/basics.subclassing.html for background information about why this is necessary)

teogale replied:
`def __array_finalize__spec(self, obj):`  as seen on AnalogSignal im using this one due to the fact that `__array_finalize__` 
doesn't let me do `self.units = getattr(obj, 'units', None)` wich is require as im creating an AnalogSignal

apdavison commented on neo/core/imagesequence.py:
```diff
@@ -0,0 +1,76 @@
+
+        return obj
+
+    def __init__(self, image_data=None, units=None, sampling_rate=None, sampling_period=None, spatial_scale=None):
```
`__init__` is not required, neither are these private variables. Because we subclass NumPy array, `self.__image_data` is the same as `self`, and `self.__sampling_rate` should just be `self.sampling_rate`, etc.

apdavison commented on neo/core/imagesequence.py:
```diff
@@ -0,0 +1,76 @@
+        analogsignal_list = []
+        for i in region_pixel:
+            data = []
+            for frame in range(len(self.__image_data)):
```
this can just be `len(self)`

apdavison commented on neo/core/imagesequence.py:
```diff
@@ -0,0 +1,76 @@
+                picture_data = 0
+                count = 0
+                for v in i:
+                    picture_data += self.__image_data[frame][v[0]][v[1]]
```
`picture_data += self.view(pq.Quantity)[frame][v[0]][v[1]]`

apdavison commented on neo/core/imagesequence.py:
```diff
@@ -0,0 +1,76 @@
+                    picture_data += self.__image_data[frame][v[0]][v[1]]
+                    count += 1
+                data.append((picture_data * 1.0) / count)
+            analogsignal_list.append(AnalogSignal(data, self.units, sampling_rate=self.__sampling_rate))
```
`self.sampling_rate`

apdavison commented on neo/core/regionofinterest.py:
```diff
@@ -0,0 +1,105 @@
+
+class CircularRegionOfInterest(RegionOfInterest):
+
+    def __init__(self, y, x, r):
```
The order of arguments should be consistent for `CircularRegionOfInterest` and `RectangularRegionOfInterest`. I suggest having `x, y` for both, rather than `y, x`

I think it would also be better to use `radius` rather than `r`

apdavison commented on neo/rawio/read_tif.py:
```diff
@@ -0,0 +1,18 @@
+from neo.core import ImageSequence
```
I've used this function as a basis to make a full IO module to go in `neo/io/tiffio.py`:
```python
 # -*- coding: utf-8 -*-
"""
Neo IO module for optical imaging data stored as a folder of TIFF images.

"""

import os
from PIL import Image
import numpy as np
from neo.core import ImageSequence, Segment, Block
from .baseio import BaseIO


class TiffIO(BaseIO):
    """
    Neo IO module for optical imaging data stored as a folder of TIFF images.
    """
    name = 'TIFF IO'
    description = "Neo IO module for optical imaging data stored as a folder of TIFF images."

    _prefered_signal_group_mode = 'group-by-same-units'
    is_readable = True
    is_writable = False

    supported_objects = [Block, Segment, ImageSequence]
    readable_objects = supported_objects
    writeable_objects = []

    support_lazy = False

    read_params = {}
    write_params = {}

    extensions = []

    mode = 'dir'

    def __init__(self, directory_path=None, **kwargs):
        BaseIO.__init__(self, directory_path, **kwargs)

    def read(self, lazy=False, units=None, sampling_rate=None, spatial_scale=None, **kwargs):
        if lazy:
            raise ValueError('This IO module does not support lazy loading')
        return [self.read_block(lazy=lazy, units=units, sampling_rate=sampling_rate,
                                spatial_scale=spatial_scale, **kwargs)]

    def read_block(self, lazy=False, units=None, sampling_rate=None,
                   spatial_scale=None, **kwargs):

        # find all the images in the given directory
        # TODO: only load files with the extension .tif or .tiff (see the glob module)
        file_name_list = os.listdir(self.filename)
        list_data_image = []
        for file_name in file_name_list:
            list_data_image.append(np.array(Image.open(self.filename + "/" + file_name), dtype=np.float))
        # todo: find out if this works for colour TIFFs, not just monochrome
        print("read block")
        image_sequence = ImageSequence(np.stack(list_data_image),
                                       units=units,
                                       sampling_rate=sampling_rate,
                                       spatial_scale=spatial_scale)
        print("creating segment")
        segment = Segment(file_origin=self.filename)
        segment.annotate(tiff_file_names=file_name_list)
        segment.imagesequences = [image_sequence]

        block = Block(file_origin=self.filename)
        segment.block = block
        block.segments.append(segment)
        print("returning block")
        return block
```

Could you do the same for `read_blk.py` --> `neo/io/blkio.py` and `read_txt.py` --> `neo/io/asciiimageio.py`?

apdavison reviewed (changes requested):
This is looking very nice. I've made some comments and suggestions below.

teogale committed:
modification of imagesequence.py and region of interest.py
(review based modification)

teogale committed:
added io file asciiimageio.py
made minor modification in the method of ImageSequence
to not have problem with quantities

teogale committed:
missing file for previous commit

teogale committed:
added  blkio.py
minor modification of asciiimageio.py

teogale committed:
added tiffio.py to myrepository file from apdavinson
modification in tiffio.py
	- load of ".tif" and ".tiff" file
	- added fonction to do natural sorting
added raise error in imagesequence

teogale committed:
- start of documentation process
- convert tif rgb into grayscale using .convert('l') /
fonction of pillow
- made change in the test_imagesequence and /
test_regionofinterest (coverage 92% and 100%)

teogale committed:
added documentation for the new class
- AsciiImageIO
- BlkIO
- TiffIO

add classes import for documention in __init_.py of io
change the way to use IO the new io classes

apdavison commented:
I notice that the Travis CI tests are failing because PIL is not installed. Please could you:
(1) edit `extras_require` in `setup.py` to indicate that `tiffio` requires PIL (or perhaps it would be better to specify `pillow`, which is a more up-to-date fork of PIL).
(2) add `pip install pillow` to the `.travis.yml` file 
(3) add `pillow` to `.circleci/requirements_testing.txt`
(4) edit `tiffio.py` so that it doesn't break Neo if PIL is not installed.

For (4), I suggest something like:
```
try:
    import PIL
    have_pil = True
except ImportError:
    have_pil = False
```

and then in `TiffIO.__init__`, add at the beginning:
```
if not have_pil:
    raise Exception("Please install the pillow package to use TiffIO")
```

teogale committed:
added test for asciiimageio.py and test_tiffio.py
added installation of pillow via setup (review based)
minor modification of asciiimageio and tiff

teogale commented:
Going in the data reported by travis-ci i got a fail from test_regionofinterest the polygon test
but when runing the test with pycharm (my editor) it doesn't fail 
i have a lot of error from dataset (ValueError: This object cannot have array annotations)

teogale committed:
modification based on review blkio and imagesequence

teogale committed:
correction of bug for test_regionofinterest

Commit:
removed array annotations from ImageSequence - need to discuss whether it makes sense, since this is essentially a single channel object; updated tests

teogale committed:
Merge pull request #1 from apdavison/imagesequence-dev

fixed tests
